File: Unidad 4-Arreglos/Ejercicio10.go
//autor: Jorge Alberto Chi leon
//Crear una matriz con numeros del 1 al n*n que no se repitan y
//verificar si lo formado es un cuadrado magico
//Entradas: Tamaño de las filas, columnas
package main

import "fmt"
import "math/rand"
import "time"

func main() {
	const size = 1000
	rand.Seed(time.Now().UnixNano())
	var n int
	for ; ;  {
	//pedimos el tamaño de la amtriz cuadrada
	fmt.Println("Dame un numero natural del 1 al 11: ")
	fmt.Scan(&n)
	if ((n<=11)&&(n>0)){
    	break;
    }
	}
    min := 1
    max := n
    //se declaran las matrices y demás variables
	var inputCuadrado [size][size]int
	var contador int
	var randomizer int
	var randomizer2 int
	var temporal int
	var suma int
	var suma2 int
	var bandera1 int
	var bandera2 int
	var anterior int
	suma = 0
	suma2 = 0
	bandera1 = 0
	bandera2 = 0
	anterior = 0

	contador = 1;
	
	//se asignan los numeros de 1 a n*n
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			inputCuadrado[i][j] = contador
			contador = contador + 1
		}
	}
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			fmt.Println("Matriz original / La posicion: [", i, "][", j, "]:", inputCuadrado[i][j])
		}
	}
	//se randomizan
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			randomizer = (rand.Intn(max - min) + min)
			randomizer2 = (rand.Intn(max - min) + min)
			temporal = inputCuadrado[i][j]
			inputCuadrado[i][j] = inputCuadrado[randomizer2][randomizer]
			inputCuadrado[randomizer2][randomizer] = temporal
		}
	}
	//se imprime la matriz ya randomizada
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			fmt.Println("Matriz randomizada / La posicion: [", i, "][", j, "]:", inputCuadrado[i][j])
		}
	}
	//se calcula si es magico (muy raro xddd) el cuadrado
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			suma = suma + inputCuadrado[i][j]
		}
		if (i > 0){
			if (suma != anterior){
				bandera1 = 0;
				i = n + 1;
			}
		}
		anterior = suma;
		suma = 0;
	}
	if (bandera1 != 0){
		bandera1 = anterior;
	}
	anterior = 0
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			suma2 = suma2 + inputCuadrado[j][i]
		}
		if (i > 0){
			if (suma2 != anterior){
				bandera2 = 0;
				i = n + 1;
			}
		}
		anterior = suma2;
		suma2 = 0;
	}
	if (bandera2 != 0){
		bandera2 = anterior;
	}
	if (bandera1 == 0){
		fmt.Println("El cuadrado no es magico...");
	}else{
		if (bandera2 == 0){
			fmt.Println("El cuadrado no es magico...");
		} else if (bandera1 == bandera2){
			fmt.Println("El cuadrado es magico...");
		} else{
			fmt.Println("El cuadrado no es magico...");
		}
	}
}

File: Unidad 4-Arreglos/Ejercicio10.c
//autor: Jorge Alberto Chi leon
//Crear una matriz con numeros del 1 al n*n que no se repitan y
//verificar si lo formado es un cuadrado magico
//Entradas: Tamaño de las filas, columnas
#include <stdlib.h>
#include <time.h>
#include <stdio.h>
void lectura(long* x);
void asignacionAleatoriaMatriz(long filaMatriz, long columnaMatriz,
	long matriz[][columnaMatriz]);
void asignacionInicialMatriz(long filaMatriz, long columnaMatriz,
	long matriz[][columnaMatriz]);
void impresionMatriz(long filaMatriz, long columnaMatriz,
	long matriz[][columnaMatriz]);
int sumaFila(long filaMatriz, long columnaMatriz,
	long matriz[][columnaMatriz]);
int sumaColumna(long filaMatriz, long columnaMatriz,
	long matriz[][columnaMatriz]);
int main()
{
	//se declaran banderas para determinar si es magico el cuadrado
	int determinadorMagicoUno = 0;
	int determinadorMagicoDos = 0;
	srand(time(0));
	long dimensiones = 0;
	//se leen las dimensiones
	lectura(&dimensiones);
	//se declara la matriz
	long matriz1alanXn[dimensiones][dimensiones];
	//se asignan y aleatorizan los numero
	asignacionInicialMatriz(dimensiones, dimensiones, matriz1alanXn);
	impresionMatriz(dimensiones, dimensiones, matriz1alanXn);
	asignacionAleatoriaMatriz(dimensiones, dimensiones, matriz1alanXn);
	printf("Matriz:\n");
	impresionMatriz(dimensiones, dimensiones, matriz1alanXn);
	//se calcula si es magico (muy raro xddd) el cuadrado
	determinadorMagicoUno = sumaFila(dimensiones, dimensiones, matriz1alanXn);
	if (determinadorMagicoUno == 0)
	{
		printf("El cuadrado no es magico...\n");
	}else{
		determinadorMagicoDos = sumaColumna(dimensiones, dimensiones,
		matriz1alanXn);
		if (determinadorMagicoDos == 0)
		{
			printf("El cuadrado no es magico...\n");
		} else if (determinadorMagicoUno == determinadorMagicoDos)
		{
			printf("El cuadrado es magico.\n");
		} else {
			printf("El cuadrado no es magico...\n");
		}
	}

	return 0;
}
void lectura(long* x){
	//se leen los datos del tamaño de los vectores
	while(1){
	printf("Dame un numero natural como maximo 11 e impar: \n");
    scanf("%d", x);
    if ((*x<=11)&&(*x>0)&&((*x % 2)!=0))
    {
    	break;
    }
	}

}
void asignacionAleatoriaMatriz(long filaMatriz, long columnaMatriz,
	//se revuelven de forma aleatorio los numeros
	long matriz[][columnaMatriz]){
	long temporal = 0;
	long randomizer = 0;
	long randomizer2 = 0;
    for (int i = 0; i < filaMatriz; ++i)
	{
		for (int j = 0; j < columnaMatriz; ++j)
		{
			randomizer = (((rand()) % filaMatriz));
			randomizer2 = (((rand()) % filaMatriz));
			temporal = matriz[i][j];
			matriz[i][j] = matriz[randomizer2][randomizer];
			matriz[randomizer2][randomizer] = temporal;
		}
	}
}
void impresionMatriz(long filaMatriz, long columnaMatriz,
	long matriz[][columnaMatriz]){
	//funcion que imprime matrices long enteras
    printf("--------\n");
    for (int i = 0; i < filaMatriz; ++i)
	{
		for (int j = 0; j < columnaMatriz; ++j)
		{
			//se imprimen los datos de manera ordenada del vector
			if (j == columnaMatriz - 1)
			{
				printf("%ld\n", matriz[i][j]);
			} else {
				printf("%ld, ", matriz[i][j]);
			}

		}
	}
	printf("--------\n");
}
void asignacionInicialMatriz(long filaMatriz, long columnaMatriz,
	//se asignan los numeros del 1 al n*n
	long matriz[][columnaMatriz]){
	long contador = 1;
	for (int i = 0; i < filaMatriz; ++i)
	{
		for (int j = 0; j < columnaMatriz; ++j)
		{
			matriz[i][j] = contador;
			contador = contador + 1;
		}
	}
}
int sumaFila(long filaMatriz, long columnaMatriz,
	//se ve si todas las filas dan el mismo numero
	long matriz[][columnaMatriz]){
	long suma = 0;
	long anterior = 0;
	int bandera = 1;
	for (int i = 0; i < filaMatriz; ++i)
	{
		for (int j = 0; j < columnaMatriz; ++j)
		{
			suma = suma + matriz[i][j];
		}
		if (i > 0)
		{
			if (suma != anterior)
			{
				bandera = 0;
				i = filaMatriz + 1;
			}
		}
		anterior = suma;
		suma = 0;
	}
	if (bandera != 0)
	{
		bandera = anterior;
	}
	return bandera;
}
int sumaColumna(long filaMatriz, long columnaMatriz,
	//se ve si todas las columnas dan el mismo numero
	long matriz[][columnaMatriz]){
	long suma = 0;
	long anterior = 0;
	int bandera = 1;
	for (int i = 0; i < filaMatriz; ++i)
	{
		for (int j = 0; j < columnaMatriz; ++j)
		{
			suma = suma + matriz[j][i];
		}
		if (i > 0)
		{
			if (suma != anterior)
			{
				bandera = 0;
				i = filaMatriz + 1;
			}
		}
		anterior = suma;
		suma = 0;
	}
	if (bandera != 0)
	{
		bandera = anterior;
	}
	return bandera;
}
